jetty-client/src/main/java/org/eclipse/jetty/client/IndexedConnectionPool.java

```diff
@@ -0,0 +1,79 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.client;
+
+import org.eclipse.jetty.client.api.Connection;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.Pool;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * <p>A {@link MultiplexConnectionPool} that picks connections at a particular
+ * index between {@code 0} and {@link #getMaxConnectionCount()}.</p>
+ * <p>The algorithm that decides the index value is decided by subclasses.</p>
+ * <p>To acquire a connection, this class obtains the index value and attempts
+ * to activate the pool entry at that index.
+ * If this activation fails, another attempt to activate an alternative pool
+ * entry is performed, to avoid stalling connection acquisition if there is
+ * an available entry at a different index.</p>
+ */
+@ManagedObject
+public abstract class IndexedConnectionPool extends MultiplexConnectionPool
+{
+    private static final Logger LOG = Log.getLogger(IndexedConnectionPool.class);
+
+    private final Pool<Connection> pool;
+
+    public IndexedConnectionPool(HttpDestination destination, int maxConnections, boolean cache, Callback requester, int maxMultiplex)
```

Comment:
I think the cache should be forced to false, otherwise you risk reaching a steady state where one thread continuously uses the connection at the same index.

Reply:
I think there shouldn't be a cache and it should be replaced by round robin: see #5218